staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	compbasemetrics "k8s.io/component-base/metrics"
+	"k8s.io/klog/v2"
+)
+
+// TimingRatioHistogram is essentially a gauge for a ratio where the client
+// independently controls the numerator and denominator.
+// When scraped it produces a histogram of samples of the ratio
+// taken at the end of every nanosecond.
+// `*TimingRatioHistogram` implements both Registerable and RatioedGauge.
+type TimingRatioHistogram struct {
+	// The implementation is layered on TimingHistogram,
+	// adding the division by an occasionally adjusted denominator.
+
+	// Registerable is the registerable aspect.
+	// That is the registerable aspect of the underlying TimingHistogram.
+	compbasemetrics.Registerable
+
+	// timingRatioHistogramInner implements the RatioedObsserver aspect.
+	timingRatioHistogramInner
+}
+
+// TimingRatioHistogramOpts is the constructor parameters of a TimingRatioHistogram.
+// The `TimingHistogramOpts.InitialValue` is the initial numerator.
+type TimingRatioHistogramOpts struct {
+	compbasemetrics.TimingHistogramOpts
+	InitialDenominator float64
+}
+
+// timingRatioHistogramInner implements the instrumentation aspect
+type timingRatioHistogramInner struct {
+	nowFunc  func() time.Time
+	delegate settable
+	sync.Mutex
+	// access only with mutex locked
+	numerator, denominator float64
+}
+
+type settable interface {
```

Comment:
I still don't understand why you need to introduce this interface (yet another one).

You create timingRatioHistogramInner in two places:
- in one delegate is TimingHistogram
- in the other it's GaugeMetric 

Both are implementing the GaugeOps interface, so why not just make a delegate be of type
`GaugeOps` instead

Reply:
No, there is a third place, and it does not use a GaugeMetric: https://github.com/MikeSpreitzer/kubernetes/blob/add-timing-ratio-histograms/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go#L211

Reply:
OK - forgot about this. But I still think we can make it easier.
My proposal is the following:

1. we change delegate to be of type `func() GaugeOps`
1. In the places I mentioned above we simply set it to `func() { return <current value of delegate> }`
1. In the place you pointed above, we change it to:
   `func() GaugeOps { return tve.delegate.WithLabelValues(tve.labelValues...) }`

This allows us to:
- get rid of this interface
- get rid of `timingHistogramVecElement` completely 

Which will simplify this code.

Reply:
done